src/backend/distributed/utils/node_metadata.c

```diff
@@ -1237,8 +1238,35 @@ static WorkerNode *
 SetNodeState(char *nodeName, int nodePort, bool isActive)
 {
 	WorkerNode *workerNode = FindWorkerNodeAnyCluster(nodeName, nodePort);
-	return SetWorkerColumn(workerNode, Anum_pg_dist_node_isactive,
-						   BoolGetDatum(isActive));
+	WorkerNode *newWorkerNode = SetWorkerColumn(workerNode,
+												Anum_pg_dist_node_isactive,
+												BoolGetDatum(isActive));
+	PropagateNodeStateMetadata(workerNode->nodeId, isActive);
```

Comment:
Hmm, this feels similar to #2320.

If we do this, we won't be able to add a new node when any node is down in the cluster. I wonder whether we should do it similar to `TriggerMetadataSync()`, which tries later and skips as long as the nodes are down.
 
I think until the metadata synced, we'd not be able to utilize the reference table shard placements on the coordinator, which is very much accepted? 

Are there anything else that'd not work until we sync `isactive `?

Reply:
> Hmm, this feels similar to #2320.

Delaying `hasmetadata` propagation wouldn't result in data inconsistency. Delaying `isactive` propagation creates a window for data inconsistency if we do DML from metadata nodes. It seems that we don't route DML to inactive nodes.

>  If we do this, we won't be able to add a new node when any node is down in the cluster. 

This is already disallowed because `master_add_node()` and `master_add_inactive_node()` send `INSERT INTO pg_dist_node` commands to workers right away.

> I think until the metadata synced, we'd not be able to utilize the reference table shard placements on the coordinator, which is very much accepted?

This is not reference table or coordinator specific. If we add a node and then create a distributed table, and then insert into it from the worker node, then I think INSERT from metadata worker won't modify the placements on the not-yet-active node.

Reply:
An example is:

```sql
-- coordinator:
select start_metadata_sync_to_node('localhost', 5433);                             
select master_activate_node('localhost', 5434);
create table t (a int, b int);
select create_distributed_table('t', 'a');

-- localhost:5433, before isactive is synced
postgres=# insert into t values (3);
ERROR:  no active placements were found for shard 102055
```

You might be fine for single-replicated tables, but imagine what happens for reference tables.

Reply:
Does master_disable_node work if a node is down? It is generally used for nodes that are down.

I agree that master_activate_node is similar to master_add_node, so requiring nodes to be up seems acceptable.

Reply:
> This is already disallowed because master_add_node() and master_add_inactive_node() send INSERT INTO pg_dist_node commands to workers right away.

Yes, sorry I missed that but thanks marco for pointing the below:

> Does master_disable_node work if a node is down?

No, it'd fail if the cluster is MX and there is at least one reference table. `master_disable_node()` -> `DeleteAllReferenceTablePlacementsFromNodeGroup()` -> `BlockWritesToShardList()` and `master_disable_node()` -> `DeleteAllReferenceTablePlacementsFromNodeGroup()`-> `SendCommandToWorkers(DELETE FROM pg_dist_placement WHERE placementid)`

I think with this PR, we're adding one more command to the above. `master_disable_node()` -> `SetNodeState()` -> `PropagateNodeStateMetadata()` -> `SendCommandToWorkers()`

@marcocitus how critical do you see that?

Reply:
Fixed this case but not doing the pg_dist_placement changes synchronously. Also added some tests.

I'm sending a meeting invite to discuss what else we could do if you don't agree with these changes.